### PR TITLE
refactor: 5x faster streaming image opto with no SSL failures

### LINKED ISSUE #1885: [bug]: 4.0.0 update, FetchError: request to graphql failed, reason: unable to verify the first certificate
**Describe the bug**
After updating to PWA Studio 4.0.0, `yarn run build` fails when verifying the certificate for the GraphQL endpoint.

**To reproduce**
1. Updated existing project to PWA Studio 4.0.0
2. `yarn install`
3. `yarn run build`

```
yarn run v1.17.3
$ yarn ce run build
$ yarn workspace @watscomagento/ce run build
$ yarn run build:prod
$ yarn run clean && yarn run load-env && webpack --no-progress --env.mode production
$ rimraf dist
$ node ../pwa-buildpack/bin/buildpack load-env --core-dev-mode .
{ FetchError: request to https://ce.test/graphql failed, reason: unable to verify the first certificate
    at ClientRequest.<anonymous> (/Users/brendan/work/clients/ce/carrierenterprise.com_m2/repo-pwa/packages/pwa-buildpack/node_modules/node-fetch/lib/index.js:1444:11)
    at ClientRequest.emit (events.js:198:13)
    at TLSSocket.socketErrorListener (_http_client.js:392:9)
    at TLSSocket.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  message:
   'request to https://ce.test/graphql failed, reason: unable to verify the first certificate',
  type: 'system',
  errno: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE',
  code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE' }
FetchError: request to https://ce.test/graphql failed, reason: unable to verify the first certificate
    at ClientRequest.<anonymous> (/Users/brendan/work/clients/ce/carrierenterprise.com_m2/repo-pwa/packages/pwa-buildpack/node_modules/node-fetch/lib/index.js:1444:11)
    at ClientRequest.emit (events.js:198:13)
    at TLSSocket.socketErrorListener (_http_client.js:392:9)
    at TLSSocket.emit (events.js:198:13)
    at emitErrorNT (internal/streams/destroy.js:91:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
    at process._tickCallback (internal/process/next_tick.js:63:19)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1
Command: /Users/brendan/.nvm/versions/node/v10.16.1/bin/node
Arguments: /usr/local/Cellar/yarn/1.17.3/libexec/lib/cli.js run build
Directory: /Users/brendan/work/clients/ce/carrierenterprise.com_m2/repo-pwa/packages/ce
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Expected behavior**
Build should finish.

**Additional context**

I had this issue back in March, but I don't know what fixed it:
https://github.com/magento/pwa-studio/pull/1045#issuecomment-475423200

Output of `curl https:/ce.test/graphql` as requested here:
https://github.com/magento/pwa-studio/issues/1052

```
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
HTTPS-proxy has similar options --proxy-cacert and --proxy-insecure.
```

**Possible solutions**

DID NOT WORK — Tried deleting the `devcert` folder as documented here.
https://manuals.gravitydept.com/frameworks/magento-pwa-studio/troubleshooting

DID NOT WORK — Tried generating the certificate as documented here:
https://magento.github.io/pwa-studio/venia-pwa-concept/setup/
`yarn buildpack create-custom-origin packages/ce`
This process succeeded, but the build does not work afterward.

DID NOT WORK — Tried deleting and regenerating the `.env` file.
`MAGENTO_BACKEND_URL="https://ce.test/" yarn buildpack create-env-file packages/ce`
This process succeeded, but the build does not work afterward.

WORKED — Changed my `.env` to use an HTTP backend URL (not HTTPS).
https://github.com/magento/pwa-studio/issues/1311
This issue suggests that fix is not needed anymore, but seems like it is.

**Please complete the following device information:**
 - Magento Version: 2.3.2
 - PWA Studio Version: 4.0.0
 - NPM version `npm -v`: 6.9.0
 - Node Version `node -v`: 10.16.1

**Please let us know what packages this bug is in regards to:**
- [ ] `venia-concept`
- [ ] `venia-ui`
- [x] `pwa-buildpack`
- [ ] `peregrine`
- [ ] `pwa-devdocs`
- [ ] `upward-js`
- [ ] `upward-spec`
- [ ] `create-pwa`

### PR DESCRIPTION
**Pull request by zetlen**
## Description

- Replace `express-sharp` with `hastily`, an Express middleware which directly simulates Fastly image optimization.
- Remove routing to image opto, since `hastily` works inline with other middlewares, modifying the response stream. Less configuration!
- Simplify `addImgOptMiddleware` a lot, since it no longer has to translate parameters AND we can deprecate some little-used features.
- Set UPWARD's `ignoreSSLErrors` configuration to respect `NODE_ENV`.

Check this out, it's super fast.

ALSO:
- Update `devcert` to include the 825-day Catalina limit.

## Related Issue

Closes #1885. Does it extra hard.
#1969 
#1311 
#1443 
#1197

### Verification Stakeholders
@magento/pwa-studio-team 

### Specification
No UPWARD spec changes, but forthcoming we might need to add custom CAs to the ProxyResolver and ServiceResolver.

### Verification Steps

1. Check out the code branch.
2. Set the MAGENTO_BACKEND_URL to the **secure (https) base url** of a local instance, or an instance with a known-bad SSL certificate.
3. Run `yarn run watch:venia`.
4. Load the homepage and click around.

Before, all the images would bomb out in a local instance because of SSL cert failures, unless you manually set very unsecure options like `NODE_TLS_REJECT_UNAUTHORIZED`. Now it doesn't, because image requests are going through the same `veniaProxy` in `packages/venia-ui/upward.yml` which already has `ignoreSSLErrors` set for the MAGENTO_BACKEND_URL origin only.

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.

### CODE REVIEW

File: packages/venia-ui/lib/components/Image/___tests__/__snapshots__/image.spec.js.snap [hunk 1]
```diff
@@ -18,7 +18,7 @@ exports[`renders an image correctly when given resource 1`] = `
     onError={[MockFunction]}
     onLoad={[MockFunction]}
     sizes="100px"
-    src="/media/catalog/product/timeless.jpg?auto=webp&format=pjpg&width=100"
+    src="/media/catalog/product/timeless.jpg?auto=webp&format=pjpg&height=undefined&width=100"
```
Comment (davemacaulay) [inline packages/venia-ui/lib/components/Image/___tests__/__snapshots__/image.spec.js.snap]: height=undefined seems a little gnarly

File: packages/venia-ui/lib/util/makeUrl.js [hunk 2]
```diff
@@ -77,32 +74,13 @@ const makeOptimizedUrl = (
...
-    /**
-     * Fit does not do anything within express-sharp, this is used within Fastly to achieve the same output of desired
-     * cropping: https://docs.fastly.com/api/imageopto/fit. Passing this to express-sharp doesn't have any side effects.
-     */
-    if (fit) {
-        params.set('fit', fit);
-    }
-
+    Object.entries(opts).forEach(([key, value]) => {
+        params.set(key, value);
```
Comment (jimbo) [inline packages/venia-ui/lib/util/makeUrl.js]:
```suggestion
        // skip null or undefined values
        if (value != null) {
          params.set(key, value);
        }
```

File: packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js [hunk 3]
```diff
@@ -11,47 +11,28 @@ try {
     markDepInvalid('apicache', e);
 }
 try {
-    expressSharp = require('@magento/express-sharp');
+    hastily = require('hastily'); // eslint-disable-line
```
Comment (jimbo) [inline packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js]: Doing this in order to throw a custom error if the require fails?

**Comment by dpatil-magento** [comment]
@zetlen  Broken image issue looks good.  But If a backend as image optimization enabled then irrespective of IMAGE_OPTIMIZING_ORIGIN=onboard or backend, images are always served by backend image optimizer (Fastly). Is this expected?

File: packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js (hunk 3 above)
Comment (zetlen) [reply packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js]: Yes!

**Pull request merged**